core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala

```diff
@@ -0,0 +1,111 @@
+  protected val executionContext: ExecutionContext
+) {
+
+  def registerGoogleProject(googleProjectReg: GoogleProjectRegistration): Future[GoogleProjectRegistration] =
+    for {
+      existingGoogleProjectReg <- googleProjectRegRepo.getGoogleProjectRegistration(googleProjectReg.googleProjectId)
+      billingProject <- billingRepository.getBillingProject(googleProjectReg.billingProjectId).map {
+        maybeBillingProject =>
+          maybeBillingProject.getOrElse(
+            throw new RawlsExceptionWithErrorReport(
+              errorReport = ErrorReport(
+                StatusCodes.NotFound,
+                "Billing project does not exist or you do not have permission to perform this action."
+              )
+            )
+          )
+      }
+      _ <- samDAO
+        .listUserActionsForResource(SamResourceTypeNames.billingProject, billingProject.projectName.value, ctx)
+        .map { actions =>
+          if (actions.isEmpty)
+            throw new RawlsExceptionWithErrorReport(
+              errorReport =
+                ErrorReport(StatusCodes.NotFound,
+                            "Billing project does not exist or you do not have permission to perform this action."
+                )
+            )
+          else if (!actions.contains(SamBillingProjectActions.link)) {
+            throw new RawlsExceptionWithErrorReport(
+              errorReport = ErrorReport(StatusCodes.Forbidden, "You do not have permission to perform this action.")
+            )
+          }
+        }
+      canLinkGoogleProject <- samDAO
+        .userHasAction(SamResourceTypeNames.googleProject,
+                       googleProjectReg.googleProjectId.value,
+                       SamGoogleProjectActions.link,
+                       ctx
+        )
+      _ = if (!canLinkGoogleProject)
+        throw new RawlsExceptionWithErrorReport(
+          errorReport = ErrorReport(StatusCodes.Forbidden, "You do not have permission to perform this action.")
+        )
+      result <- existingGoogleProjectReg match {
+        case Some(existing) if existing.billingProjectId == googleProjectReg.billingProjectId =>
+          Future.successful(existing)
+        case Some(existing) if existing.billingProjectId != googleProjectReg.billingProjectId =>
+          throw new RawlsExceptionWithErrorReport(
+            errorReport = ErrorReport(StatusCodes.Conflict,
+                                      "This google project id is already registered with a different billing project."
+            )
+          )
+        case _ =>
+          val updatedGoogleProjectReg = googleProjectReg.copy(billingAccount = billingProject.billingAccount)
+          for {
+            _ <- googleServicesDAO
+              .setBillingAccountName(googleProjectReg.googleProjectId,
+                                     billingProject.billingAccount.get,
+                                     ctx.toTracingContext
+              )
+              .recoverWith { case ex: Exception =>
```

Comment:
Struggling with scala again - I've tried a few variations on `recover`, `recoverWith` etc, and yet in the test it doesn't seem to have wrapped it in a RawlsException.

Reply:
Hm, looking more carefully at the `GoogleServicesDao`, I think that if there's a problem setting the billing account then that will throw the error and I'm not sure we need to address it specifically in the GPRService.